db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/FromSubquery.java

```diff
@@ -310,14 +312,26 @@ public void bindExpressions(FromList fromListParam)
          * the table since the view was created.
          */
         subqueryRCL = subquery.getResultColumns();
-        if (resultColumns != null && resultColumns.getCountMismatchAllowed() &&
-            resultColumns.size() < subqueryRCL.size())
-        {
-            for (int index = subqueryRCL.size() - 1;
-                 index >= resultColumns.size();
-                 index--)
-            {
-                subqueryRCL.removeElementAt(index);
+        if (resultColumns != null && resultColumns.getCountMismatchAllowed()) {
+            if (resultColumns.size() < subqueryRCL.size()) {
+                // columns added to underlying table
+                for (int index = subqueryRCL.size() - 1;
+                     index >= resultColumns.size();
+                     index--) {
+                    subqueryRCL.removeElementAt(index);
+                }
+            } else if (resultColumns.size() > subqueryRCL.size()) {
+                // columns dropped from underlying table
+                throw StandardException.newException(SQLState.LANG_DERIVED_COLUMN_LIST_MISMATCH, correlationName);
+            } else {
+                // number of columns is OK, check column names
+                for (int i = 0; i < resultColumns.size(); ++i) {
+                    ResultColumn thisRC = resultColumns.elementAt(i);
+                    ResultColumn subqRC = subqueryRCL.elementAt(i);
+                    if (thisRC.getName() != null && !thisRC.getName().equals(subqRC.getName())) {
```

Comment:
Having `thisRC.getName() != null` is mostly defensive programming since we call `.equals()` on it. Usually for an expression like `col1 / 3.14` we should generate an internal name during planning, something like `SQLCol1`. On the other side, passing `null` into `String.equals()` should be fine.